File: num2binaries.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void binary(int num) {
    int result[32], i = 0;
    int mask = 1;
    while (num > 0) {
        if (num & mask) {
            result[i] = mask;
            i++;
            num -= mask;
        }
        mask <<= 1;
    }
    for (int j = i - 1; j >= 0; j--) {
        printf("%d ", result[j]);
    }
    printf("\n");
}

int main(int argc, char *argv[]) {
    char *file_name = NULL;
    FILE *fp;

    if (argc > 2 && strcmp(argv[1], "-f") == 0) {
        file_name = argv[2];
        fp = fopen(file_name, "r");
        if (fp == NULL) {
            perror("Error: Could not open file");
            exit(EXIT_FAILURE);
        }
    } else {
        fp = stdin;
    }

    char *line = NULL;
    size_t len = 0;
    ssize_t read;

    while ((read = getline(&line, &len, fp)) != -1) {
        line[strcspn(line, "\n")] = 0; // remove newline character
        char *endptr;
        long num = strtol(line, &endptr, 10);
        if (*endptr == '\0') {
            binary(num);
        } else {
            fprintf(stderr, "Warning: %s is not a number\n", line);
        }
    }

    if (file_name) {
        fclose(fp);
    }

    return 0;
}

File: num2binaries.php
<?php

function binary($num) {
    $result = array();
    $mask = 1;
    while ($num > 0) {
        if ($num & $mask) {
            $result[] = $mask;
            $num -= $mask;
        }
        $mask <<= 1;
    }
    echo implode(" ", $result) . PHP_EOL;
}

$options = getopt("f:");

if (isset($options['f'])) {
    $file = fopen($options['f'], 'r');
    if ($file === false) {
        echo "Error: Could not open file" . PHP_EOL;
        exit(1);
    }
    while (($line = fgets($file)) !== false) {
        $line = trim($line);
        if (is_numeric($line)) {
            $num = (int)$line;
            binary($num);
        } else {
            fprintf(STDERR, "Warning: %s is not a number\n", $line);
        }
    }
    fclose($file);
} else {
    while (($line = fgets(STDIN)) !== false) {
        $line = trim($line);
        if (is_numeric($line)) {
            $num = (int)$line;
            binary($num);
        } else {
            fprintf(STDERR, "Warning: %s is not a number\n", $line);
        }
    }
}

